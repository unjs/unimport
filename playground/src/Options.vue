<script lang="ts">
export default defineComponent({
  setup () {
    const multiplier = ref(100)
    const count = ref(5)

    return {
      multiplier,
      count,
      inc () {
        count.value += 1
      },
      bump () {
        multiplier.value += 1
      },
      nested () {
        return 'from setup, this should not be overridden'
      }
    }
  }
})
</script>

<template>
  <div>
    <h1>{{ count }} x {{ multiplier }} = {{ count * multiplier }}</h1>
    <button @click="inc">
      Inc
    </button>
    <button @click="bump">
      x1
    </button>
    <div>
      {{ nested() }}
    </div>
  </div>
</template>
