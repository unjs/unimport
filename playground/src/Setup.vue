<script setup lang="ts">
const count: Ref<number> = ref(1)

function inc () {
  count.value += 1
}
</script>

<template>
  <div>
    <h1>{{ count }} x {{ multiplier }} = {{ count * multiplier }}</h1>
    <button @click="inc">
      Inc
    </button>
    <button @click="bump">
      x1
    </button>
    <div>
      {{ nested() }}
    </div>
  </div>
</template>
