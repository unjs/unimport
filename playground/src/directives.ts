import type { DirectiveBinding } from 'vue'

function mounted(el: HTMLElement, binding: DirectiveBinding) {
  console.log('mounted', el, binding)
}

function unmounted(el: HTMLElement, binding: DirectiveBinding) {
  console.log('unmounted', el, binding)
}

function updated(el: HTMLElement, binding: DirectiveBinding) {
  console.log('updated', el, binding)
}

export const CustomDirective = {
  mounted,
  unmounted,
  updated,
}

export function AwesomeDirective(el: HTMLElement, binding: DirectiveBinding) {
  console.log('AwesomeDirective', el, binding)
}
