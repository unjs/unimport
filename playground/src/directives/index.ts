export { AwesomeDirective } from './awesome-directive'
export { CustomDirective } from './custom-directive'
