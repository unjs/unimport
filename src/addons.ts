export { vueDirectivesAddon } from './addons/vue-directives'
export { vueTemplateAddon } from './addons/vue-template'
