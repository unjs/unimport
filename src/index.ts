import { detectSyntax } from 'mlly'
import escapeRE from 'escape-string-regexp'
import type { Import, UnimportOptions } from './types'
import { excludeRE, stripCommentsAndStrings, toImports, separatorRE, importAsRE } from './utils'
import { resolvePreset } from './preset'
export * from './types'

interface Context {
  imports: Import[]
  matchRE: RegExp
  map: Map<string, Import>
}

export interface Unimport {
  addImports: (code: string) => ReturnType<typeof addImports>
  detectImports: (code: string) => ReturnType<typeof detectImports>
}

export function createUnimport (opts: Partial<UnimportOptions>): Unimport {
  const ctx: Context = {
    imports: [].concat(opts.imports).filter(Boolean),
    map: new Map(),
    matchRE: /__never__/g
  }

  // Resolve presets
  for (const preset of opts.presets || []) {
    ctx.imports.push(...resolvePreset(preset))
  }

  // Normalize imports
  for (const _import of ctx.imports) {
    _import.as = _import.as || _import.name
  }

  // Detect duplicates
  const usedNames = new Set()
  for (const autoImport of ctx.imports) {
    if (!usedNames.has(autoImport.as)) {
      usedNames.add(autoImport.as)
    }
  }

  // Create regex
  ctx.matchRE = new RegExp(`(?:\\b|^)(${ctx.imports.map(i => escapeRE(i.as)).join('|')})(?:\\b|\\()`, 'g')

  // Create map
  for (const _import of ctx.imports) {
    ctx.map.set(_import.as, _import)
  }

  return {
    detectImports: (code: string) => detectImports(code, ctx),
    addImports: (code: string) => addImports(code, ctx)
  }
}

function detectImports (code: string, ctx: Context) {
  // Strip comments so we don't match on them
  const strippedCode = stripCommentsAndStrings(code)
  const isCJSContext = detectSyntax(strippedCode).hasCJS

  // Find all possible injection
  const matched = new Set(
    Array.from(strippedCode.matchAll(ctx.matchRE)).map(i => i[1])
  )

  // Remove those already defined
  for (const regex of excludeRE) {
    Array.from(strippedCode.matchAll(regex))
      .flatMap(i => [
        ...(i[1]?.split(separatorRE) || []),
        ...(i[2]?.split(separatorRE) || [])
      ])
      .map(i => i.replace(importAsRE, '').trim())
      .forEach(i => matched.delete(i))
  }

  const matchedImports = Array.from(matched)
    .map(name => ctx.map.get(name))
    .filter(Boolean)

  return {
    strippedCode,
    isCJSContext,
    matchedImports
  }
}

function addImports (code: string, ctx: Context) {
  const { isCJSContext, matchedImports } = detectImports(code, ctx)
  if (!matchedImports.length) {
    return { code }
  }
  const imports = toImports(matchedImports, isCJSContext)
  return { code: imports + code }
}
