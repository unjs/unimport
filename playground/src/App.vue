<script setup lang="ts">
import Setup from './Setup.vue'
import Options from './Options.vue'
</script>

<template>
  <div>
    <h1>Setup API (auto imported)</h1>
    <Setup />
    <h1>Options API</h1>
    <Options />
  </div>
</template>

<style>
#app {
  font-family: Avenir, Helvetica, Arial, sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  text-align: center;
  color: #2c3e50;
  margin-top: 60px;
}
</style>
