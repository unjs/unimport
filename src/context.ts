import { detectSyntax } from 'mlly'
import MagicString from 'magic-string'
import type { Addon, Import, InjectImportsOptions, Thenable, TypeDeclarationOptions, UnimportContext, UnimportOptions } from './types'
import { excludeRE, stripCommentsAndStrings, separatorRE, importAsRE, toTypeDeclarationFile, addImportToCode, dedupeImports, toExports, normalizeImports, matchRE, getMagicString, getString } from './utils'
import { resolveBuiltinPresets } from './preset'
import { vueTemplateAddon } from './addons'

export type Unimport = ReturnType<typeof createUnimport>

export function createUnimport (opts: Partial<UnimportOptions>) {
  // Cache for combine imports
  let _combinedImports: Import[] | undefined
  const _map = new Map()

  const addons: Addon[] = []

  if (Array.isArray(opts.addons)) {
    addons.push(...opts.addons)
  } else if (opts.addons?.vueTemplate) {
    addons.push(vueTemplateAddon())
  }

  const ctx: UnimportContext = {
    staticImports: [...(opts.imports || [])].filter(Boolean),
    dynamicImports: [],
    get imports () {
      return updateImports()
    },
    get map () {
      updateImports()
      return _map
    },
    invalidate () {
      _combinedImports = undefined
    },
    resolveId: (id, parentId) => opts.resolveId?.(id, parentId),
    addons
  }

  // Resolve presets
  ctx.staticImports.push(...resolveBuiltinPresets(opts.presets || []))

  function updateImports () {
    if (!_combinedImports) {
      // Combine static and dynamic imports
      // eslint-disable-next-line no-console
      const imports = normalizeImports(dedupeImports([...ctx.staticImports, ...ctx.dynamicImports], opts.warn || console.warn))
        .filter(i => !i.disabled)

      // Create map
      _map.clear()
      for (const _import of imports) {
        _map.set(_import.as ?? _import.name, _import)
      }

      _combinedImports = imports
    }
    return _combinedImports
  }

  async function modifyDynamicImports (fn: (imports: Import[]) => Thenable<void | Import[]>) {
    const result = await fn(ctx.dynamicImports)
    if (Array.isArray(result)) {
      ctx.dynamicImports = result
    }
    ctx.invalidate()
  }

  function clearDynamicImports () {
    ctx.dynamicImports.length = 0
    ctx.invalidate()
  }

  function generateTypeDeclarations (options?: TypeDeclarationOptions) {
    const opts: TypeDeclarationOptions = {
      resolvePath: i => i.from.replace(/\.ts$/, ''),
      ...options
    }
    let dts = toTypeDeclarationFile(ctx.imports, opts)
    for (const addon of ctx.addons) {
      dts = addon.declaration?.call(ctx, dts, opts) ?? dts
    }
    return dts
  }

  updateImports()

  return {
    clearDynamicImports,
    modifyDynamicImports,
    getImports: () => ctx.imports,
    detectImports: (code: string) => detectImports(code, ctx),
    injectImports: (code: string | MagicString, id?: string, options?: InjectImportsOptions) => injectImports(code, id, ctx, options),
    toExports: () => toExports(ctx.imports),
    generateTypeDeclarations
  }
}

// eslint-disable-next-line require-await
async function detectImports (code: string | MagicString, ctx: UnimportContext) {
  // Strip comments so we don't match on them
  const strippedCode = stripCommentsAndStrings(getString(code))
  const syntax = detectSyntax(strippedCode)
  const isCJSContext = syntax.hasCJS && !syntax.hasESM

  // Find all possible injection
  const identifiers = new Set(
    Array.from(strippedCode.matchAll(matchRE))
      .map(i => i[1] === '.' ? '' : i[2]) // Remove dot access
      .filter(Boolean)
  )

  // Remove those already defined
  for (const regex of excludeRE) {
    Array.from(strippedCode.matchAll(regex))
      .flatMap(i => [
        ...(i[1]?.split(separatorRE) || []),
        ...(i[2]?.split(separatorRE) || [])
      ])
      .map(i => i.replace(importAsRE, '').trim())
      .forEach(i => identifiers.delete(i))
  }

  let matchedImports = Array.from(identifiers)
    .map(name => ctx.map.get(name))
    .filter(i => i && !i.disabled) as Import[]

  for (const addon of ctx.addons) {
    matchedImports = await addon.matchImports?.call(ctx, identifiers, matchedImports) || matchedImports
  }

  return {
    strippedCode,
    isCJSContext,
    matchedImports
  }
}

async function injectImports (code: string | MagicString, id: string | undefined, ctx: UnimportContext, options?: InjectImportsOptions) {
  const s = getMagicString(code)

  for (const addon of ctx.addons) {
    await addon.transform?.call(ctx, s, id)
  }

  const { isCJSContext, matchedImports } = await detectImports(s, ctx)
  const imports = await resolveImports(ctx, matchedImports, id)

  return addImportToCode(s, imports, isCJSContext, options?.mergeExisting)
}

async function resolveImports (ctx: UnimportContext, imports: Import[], id: string | undefined) {
  const resolveCache = new Map<string, string>()

  return await Promise.all(imports.map(async (i) => {
    if (!resolveCache.has(i.from)) {
      resolveCache.set(i.from, await ctx.resolveId(i.from, id) || i.from)
    }
    return <Import>{
      ...i,
      from: resolveCache.get(i.from)!
    }
  }))
}
