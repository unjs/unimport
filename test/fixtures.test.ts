/// <reference types="vite/client" />

import { basename } from 'path'
import { describe, it, expect } from 'vitest'
import { createUnimport } from '../src/context'

const UNMODIFIED_MARK = '@test-unmodified'

describe('fixtures', () => {
  const { injectImports } = createUnimport({
    imports: [
      { name: 'default', from: 'default', as: 'customDefault' },
      { name: 'foobar', from: 'foobar', as: 'foobar' },
      { name: 'fooConst', from: 'foo', as: 'fooConst' },
      { name: 'test', from: 'vitest', as: 'test' },
      { name: 'Foo', from: 'foobar', as: 'Foo' },
      { name: 'Bar', from: 'foobar', as: 'Bar' },
      { name: 'it', from: 'it', as: 'it' },
      { name: 'fit', from: 'fit', as: 'fit' }
    ],
    presets: [
      {
        from: 'vue',
        imports: ['ref', 'reactive', 'computed', 'toRefs']
      },
      {
        from: 'three',
        imports: [['*', 'THREE']]
      },
      {
        from: 'react',
        imports: ['useState', 'useEffect', 'useRef']
      },
      {
        from: 'jquery',
        imports: ['$']
      }
    ]
  })

  const files = import.meta.glob('./fixtures/*', {
    as: 'raw',
    eager: true
  })

  for (const [file, code] of Object.entries(files)) {
    it(basename(file), async () => {
      const pass1 = (await injectImports(code))?.code ?? code
      if (code.includes(UNMODIFIED_MARK)) {
        expect(pass1).toBe(code)
      } else {
        expect(pass1).toMatchSnapshot()
        expect(pass1).not.toEqual(code)
        const pass2 = (await injectImports(pass1))?.code ?? pass1
        expect(pass2).toBe(pass1)
      }
    })
  }
})
