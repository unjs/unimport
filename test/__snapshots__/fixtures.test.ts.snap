// Vitest Snapshot v1

exports[`fixtures > comments.ts 1`] = `
"import { ref, computed, toRefs } from 'vue';
import { useEffect } from 'react';
/* import { ref } from 'vue' */
// import { computed } from 'vue'

const a = {
  rel: \\"noreferrer\\",
  href: \\"https://github.com/antfu/vitesse\\",
  target: \\"_blank\\"
}

const foo = ref(0)
const bar = computed(() => {})

const b = { class: \\"text-sm opacity-75\\" };
const templateLiteralWithDoubleSlash = \`http://\${useEffect()}\`
const templateLiteralWithPseudoComment = \`/* \${toRefs()} */\`
// reactive
//reactive
/*reactive*/
/**/
"
`;

exports[`fixtures > custom.js 1`] = `
"import customDefault from 'default';
const a = customDefault(customNamed())
const a = customDefaultAlias()
"
`;

exports[`fixtures > dollar.js 1`] = `
"import { $ } from 'jquery';
$('hi')
"
`;

exports[`fixtures > existing6.jsx 1`] = `
"import { computed } from 'vue';
const a = computed(() => /^(https?:\\\\/\\\\/|\\\\/\\\\/)/.test(props.to))
"
`;

exports[`fixtures > existing8.ts 1`] = `
"import { bar } from 'specifier2'
export { bar }
import foo from 'specifier1' // error
export { foo }
"
`;

exports[`fixtures > export-default.ts 1`] = `
"import { ref } from 'vue';
export default ref(
  1
)
"
`;

exports[`fixtures > false-regex.ts 1`] = `
"import { ref } from 'vue';
// https://github.com/antfu/unplugin-auto-import/issues/220
const value = 1 / 1 + ref().val / 1
console.log(value)
"
`;

exports[`fixtures > function-call.ts 1`] = `
"import { ref } from 'vue';
useFoo(ref, 1)
"
`;

exports[`fixtures > glob.ts 1`] = `
"import { ref } from 'vue';
const modules = import.meta.glob(\\"/src/nested/*.vue\\");
const msg = ref(\\"Global Imports\\");
"
`;

exports[`fixtures > import-all.ts 1`] = `
"import * as THREE from 'three';
const scene = new THREE.Scene()
"
`;

exports[`fixtures > in.js 1`] = `
"import { Foo, Bar } from 'foobar';
Foo instanceof Bar
"
`;

exports[`fixtures > instanceof.js 1`] = `
"import { Foo, Bar } from 'foobar';
// Foo instanceof Bar
Foo instanceof Bar
"
`;

exports[`fixtures > no-eol.js 1`] = `
"import { Foo } from 'foobar';
Foo"
`;

exports[`fixtures > props.ts 1`] = `
"import { toRefs } from 'vue';
const a = z.ref(0)
console.log({
  ...toRefs(a)
})
"
`;

exports[`fixtures > prototype.ts 1`] = `
"import { foobar } from 'foobar';
const a = Object.keys(foobar)
const c = a.toString()
"
`;

exports[`fixtures > react.jsx 1`] = `
"import { useState } from 'react';
export function Component() {
  const [count, setCount] = useState()

  return <div>{ count }</div>
}
"
`;

exports[`fixtures > switch.js 1`] = `
"import { Foo, Bar } from 'foobar';
switch (mode) {
  case Foo:
    return '1'
  case Bar + 1:
    return '2'
}
"
`;

exports[`fixtures > template-literal.ts 1`] = `
"import { ref, reactive, computed, toRefs } from 'vue';
import { fooConst } from 'foo';
import { $ } from 'jquery';
import { useState, useEffect, useRef } from 'react';
const z = \`bar-\${
  ref()
}-\${fooConst}\`
const withRegex = /\`/
const secondLine = [$, \`\`]
const nestedSingleQuotes = \`'\${reactive()}'\`
const nestedDoubleQuotes = \`\\"\${computed()}\\"\`
const nestedTemplate = \`prefix1\${\`prefix2\${toRefs()}\`}\`
const multilineTemplate = \`\${useState(
)}\`

const multilineNestedTemplateAndQuotes = \`\\"'\${useEffect(\`'\${
useRef(
)
}'\`)}\\"'\`
"
`;

exports[`fixtures > template-ref.vue 1`] = `
"import { ref } from 'vue';
<template>
  <div ref=\\"divRef\\"></div>
</template>

<script setup>
const divRef = ref(null)
</script>
"
`;

exports[`fixtures > template-tag.ts 1`] = `
"import { ref } from 'vue';
const z = ref\`barfoo\`
"
`;

exports[`fixtures > ternary.js 1`] = `
"import { Foo, Bar } from 'foobar';
Foo instanceof Bar ? true : false
"
`;

exports[`fixtures > ts-generic.ts 1`] = `
"import { useState } from 'react';
useState<number>()
"
`;

exports[`fixtures > vue.js 1`] = `
"import { ref } from 'vue';
const a = ref(0)
"
`;
